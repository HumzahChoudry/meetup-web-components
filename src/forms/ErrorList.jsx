import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

const ERROR_CLASSNAME = 'text--error text--small';

/**
 * ErrorList
 *
 * Accessible list of errors for a form input.
 *
 * Currently supports only a single error due to a
 * limitation with `redux-form`.
 */
const ErrorList = props => {
	const { errorId, error, ...other } = props;

	if (errorId) {
		other.id = errorId;
	}

	return (
		<ul aria-live="assertive" {...other}>
			{error && <li className={cx(ERROR_CLASSNAME, 'text--small')}>{error}</li>}
		</ul>
	);
};
ErrorList.propTypes = {
	errorId: PropTypes.string,
	error: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element,
		PropTypes.bool,
	]),
};

export default ErrorList;
