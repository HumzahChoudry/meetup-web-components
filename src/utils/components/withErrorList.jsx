import React from 'react';
import ErrorList from '../../forms/ErrorList';

/**
 * @function getErrorId
 * returns an id with which we can set ARIA
 * relationships between errors and form fields
 *
 * @param {string} id - the id of the form field
 * @returns {string} id for error list
 */
export const getErrorId = (id) => `${id}-error`;


/**
 * Provides `ErrorList` component and aria related props
 * to wrapped form component
 *
 * @param {React.element} WrappedComponent - the component to wrap
 */
const withErrorList = WrappedComponent => {
	/**
	 * @module WithErrorList
	 */
	class WithErrorList extends React.PureComponent {
		/**
		 * @constructor
		 * @param {Object} - React element props
		 */
		constructor(props) {
			super(props);
		}

		/**
		 * Adds props to the wrapped form component for a11y.
		 * Adds `ErrorList` as adjacent sibling to wrapped form.
		 *
		 * @returns {React.element}
		 */
		render() {
			const {
				id,
				error,
			} = this.props;

			const fieldProps = {
				id,
				'aria-invalid': Boolean(error),
				'aria-describedby': id && getErrorId(id),
			};
			const listProps = {
				errorId: id && getErrorId(id),
				errors: error,
			};

			return this.props.suppressError ? <WrappedComponent /> : (
				<div>
					<WrappedComponent {...this.props} {...fieldProps} />
					<ErrorList {...listProps} />
				</div>
			);
		}
	}

	const wrappedComponentName =
		WrappedComponent.displayName || WrappedComponent.name || 'Component';

	WithErrorList.WrappedComponent = WrappedComponent;
	WithErrorList.displayName = `WithErrorList(${wrappedComponentName})`;

	return WithErrorList;
};

export default withErrorList;
