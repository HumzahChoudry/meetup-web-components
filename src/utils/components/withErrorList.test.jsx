import React from 'react';
import withErrorList, { getErrorId } from './withErrorList';
import ErrorList from '../../forms/ErrorList';
import { shallow } from 'enzyme';

const MOCK_FIELD_ID = 'name';
const MOCK_ERROR_MSG = 'You are not Keith Hernandez';

const TestComponent = (props) => (<h1 {...props}>Hello world</h1>);
const TestComponentWithErrorList = withErrorList(TestComponent);

describe('error id', () => {
	it('correctly generates id for `aria-describedby` from field `id` prop', () => {
		const actual = getErrorId(MOCK_FIELD_ID);
		const expected = 'name-error';
		expect(actual).toBe(expected);
	});
});

describe('withErrorList wrapped component', () => {
	const component = shallow(
		<TestComponentWithErrorList
			id={MOCK_FIELD_ID}
			error={MOCK_ERROR_MSG}
		/>
	);
	const componentNoError = shallow(
		<TestComponentWithErrorList
			id={MOCK_FIELD_ID}
		/>
	);
	const componentNoIdNoError = shallow(
		<TestComponentWithErrorList />
	);

	it('matches snapshot for component with id and error props', () => {
		expect(component).toMatchSnapshot();
	});
	it('matches snapshot for component without error prop', () => {
		expect(componentNoError).toMatchSnapshot();
	});
	it('matches snapshot for component without id or error props', () => {
		expect(componentNoIdNoError).toMatchSnapshot();
	});

	it('passes `error` to the `ErrorList`', () => {
		const errorProp = component.find(ErrorList).prop('error');
		expect(errorProp).toBe(MOCK_ERROR_MSG);
	});

	it('provides id for `aria-describedby` in both wrapped components when id is provided', () => {
		const wrappedComponent = componentNoError.find(TestComponent);
		const errorList = component.find(ErrorList);
		const expected = getErrorId(MOCK_FIELD_ID);

		expect(wrappedComponent.prop('aria-describedby'))
			.toEqual(expected);
		expect(errorList.prop('id'))
			.toEqual(expected);
	});

	it('does NOT provide id for `aria-describedby` in both wrapped components when id is NOT provided', () => {
		const wrappedComponent = componentNoIdNoError.find(TestComponent);
		const errorList = component.find(ErrorList);

		expect(wrappedComponent.prop('aria-describedby')).toBe(undefined);
		expect(errorList.prop('errorId')).toBe(undefined);
	});

	it('provides a wrapped displayname', () => {
		expect(TestComponentWithErrorList.displayName).toBe(
			'WithErrorList(TestComponent)'
		);
		expect(true).toBe(true);
	});
});
