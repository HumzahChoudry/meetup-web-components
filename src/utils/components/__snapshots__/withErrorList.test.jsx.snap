// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withErrorList wrapped component matches snapshot for component with id and error props 1`] = `
<div>
  <TestComponent
    aria-describedby="name-error"
    aria-invalid={true}
    error="You are not Keith Hernandez"
    id="name"
  />
  <ErrorList
    error="You are not Keith Hernandez"
    id="name-error"
  />
</div>
`;

exports[`withErrorList wrapped component matches snapshot for component without error prop 1`] = `
<div>
  <TestComponent
    aria-describedby="name-error"
    aria-invalid={false}
    id="name"
  />
  <ErrorList
    id="name-error"
  />
</div>
`;

exports[`withErrorList wrapped component matches snapshot for component without id or error props 1`] = `
<div>
  <TestComponent
    aria-invalid={false}
  />
  <ErrorList />
</div>
`;
